packages/keyring-controller/src/KeyringController.ts

```diff
@@ -218,6 +220,17 @@ export class KeyringController extends BaseController<
       throw new Error('No HD keyring found');
     }
     const oldAccounts = await this.#keyring.getAccounts();
+
+    if (oldAccounts.length !== accountCount) {
```

Comment:
How should we handle the case where `oldAccounts.length` is _below_ `accountCount`? In that case, we'll be generating the wrong account.

Perhaps this method should throw an error if we detect there are fewer accounts than the caller expects? This would make it non-idempotent in that error case, but I don't see another option. Aside maybe from.... generating multiple accounts if necessary, but that would be unexpected for the caller as well. Hmm....

Reply:
We can apply [this suggestion](https://github.com/MetaMask/core/pull/1298#discussion_r1182559638) and throw an error if we detect a call out of sequence, eg:
```typescript
addNewAccount(10); // returns 11th account
addNewAccount(12); // throws error
addNewAccount(11); // returns 12th account
```

Otherwise, we would be returning `addedAccountAddress: undefined` as `accounts[12]` doesn't exist in that example case

Reply:
That sounds good to me